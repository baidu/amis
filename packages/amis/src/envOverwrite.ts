/**
 * @file 用于在移动端或不同语言环境下使用不同配置
 */

import {SchemaNode, Schema} from './types';
import {RendererProps, RendererConfig, addSchemaFilter} from './factory';
import {findObjectsWithKey} from './utils/helper';
const isMobile = (window as any).matchMedia?.('(max-width: 768px)').matches
  ? true
  : false;

// 这里不能用 addSchemaFilter 是因为还需要更深层的替换，比如 select 里的 options
export const envOverwrite = (schema: any, locale?: string) => {
  if (schema.mobile && isMobile) {
    Object.assign(schema, schema.mobile);
    delete schema.mobile;
  }

  if (locale) {
    let schemaNodes = findObjectsWithKey(schema, locale);
    for (let schemaNode of schemaNodes) {
      Object.assign(schemaNode, schemaNode[locale]);
      delete schemaNode[locale];
    }
  }

  if (isMobile) {
    let schemaNodes = findObjectsWithKey(schema, 'mobile');
    for (let schemaNode of schemaNodes) {
      Object.assign(schemaNode, schemaNode['mobile']);
      delete schemaNode['mobile'];
    }
  }
};
