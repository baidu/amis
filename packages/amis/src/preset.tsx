import {
  addRootWrapper,
  extendDefaultEnv,
  LazyComponent,
  render,
  themeable,
  ThemeProps
} from 'amis-core';

import {alert, confirm, setRenderSchemaFn, toast, ImageGallery} from 'amis-ui';
import React from 'react';

extendDefaultEnv({
  alert,
  confirm,
  notify: (type: keyof typeof toast, msg: string, conf: any) =>
    toast[type] ? toast[type](msg, conf) : console.warn('[Notify]', type, msg)
});

setRenderSchemaFn((controls, value, callback, scopeRef, theme) => {
  return render(
    {
      name: 'form',
      type: 'form',
      wrapWithPanel: false,
      mode: 'horizontal',
      controls,
      messages: {
        validateFailed: ''
      }
    },
    {
      data: value,
      onFinished: callback,
      scopeRef,
      theme
    },
    {
      session: 'prompt'
    }
  );
});

addRootWrapper((props: any) => {
  const {env, children} = props;
  return (
    <ImageGallery modalContainer={env.getModalContainer}>
      {children}
    </ImageGallery>
  );
});

const SimpleSpinner = themeable(
  (
    props: {
      className?: string;
      spinnerClassName?: string;
    } & ThemeProps
  ) => {
    const cx = props.classnames;

    return (
      <div
        data-testid="spinner"
        className={cx(`Spinner`, 'in', props.className)}
      >
        <div
          className={cx(
            `Spinner-icon`,
            'Spinner-icon--default',
            props.spinnerClassName
          )}
        ></div>
      </div>
    );
  }
);

LazyComponent.defaultProps.placeholder = <SimpleSpinner />;
