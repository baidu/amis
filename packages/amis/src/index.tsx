/** @license amis v@version
 *
 * Copyright Baidu
 *
 * This source code is licensed under the Apache license found in the
 * LICENSE file in the root directory of this source tree.
 */
export * from 'amis-core';
export * from 'amis-ui';

import './preset';

// 注册渲染器
import './renderers/Action';
import './renderers/Alert';
import './renderers/App';
import './renderers/Avatar';
import './renderers/Remark';
import './renderers/ButtonGroup';
import './renderers/Form/ButtonToolbar';
import './renderers/Breadcrumb';
import './renderers/DropDownButton';
import './renderers/Calendar';
import './renderers/Collapse';
import './renderers/CollapseGroup';
import './renderers/Color';
import './renderers/CRUD';
import './renderers/CRUD2';
import './renderers/Pagination';
import './renderers/Cards';
import './renderers/Card';
import './renderers/Card2';
import './renderers/Custom';
import './renderers/Date';
import './renderers/Dialog';
import './renderers/Divider';
import './renderers/Each';
import './renderers/Flex';
import './renderers/Form/Control';
import './renderers/Form/Hidden';
import './renderers/Form/InputText';
import './renderers/Form/InputTag';
import './renderers/Form/InputNumber';
import './renderers/Form/Textarea';
import './renderers/Form/Checkboxes';
import './renderers/Form/Checkbox';
import './renderers/Form/InputCity';
import './renderers/Form/ChartRadios';
import './renderers/Form/InputRating';
import './renderers/Form/Switch';
import './renderers/Form/Radios';
import './renderers/Form/Radio';
import './renderers/Form/JSONSchema';
import './renderers/Form/JSONSchemaEditor';
import './renderers/Form/ListSelect';
import './renderers/Form/LocationPicker';
import './renderers/Form/Select';
import './renderers/Form/Static';
import './renderers/Form/InputDate';
import './renderers/Form/InputDateRange';
import './renderers/Form/InputFormula';
import './renderers/Form/InputRepeat';
import './renderers/Form/InputTree';
import './renderers/Form/TreeSelect';
import './renderers/Form/InputImage';
import './renderers/Form/InputFile';
import './renderers/Form/UUID';
import './renderers/Form/MatrixCheckboxes';
import './renderers/Form/InputMonthRange';
import './renderers/Form/InputQuarterRange';
import './renderers/Form/InputYearRange';
import './renderers/Form/InputRange';
import './renderers/Form/InputArray';
import './renderers/Form/Combo';
import './renderers/Form/ConditionBuilder';
import './renderers/Form/InputSubForm';
import './renderers/Form/InputExcel';
import './renderers/Form/InputRichText';
import './renderers/Form/Editor';
import './renderers/Form/DiffEditor';
import './renderers/Form/InputColor';
import './renderers/Form/ChainedSelect';
import './renderers/Form/NestedSelect';
import './renderers/Form/Transfer';
import './renderers/Form/TransferPicker';
import './renderers/Form/InputTable';
import './renderers/Form/Picker';
import './renderers/Form/IconPicker';
import './renderers/Form/IconSelect';
import './renderers/Form/Formula';
import './renderers/Form/FieldSet';
import './renderers/Form/TabsTransfer';
import './renderers/Form/TabsTransferPicker';
import './renderers/Form/Group';
import './renderers/Form/InputGroup';
import './renderers/Form/UserSelect';
import './renderers/Grid';
import './renderers/Grid2D';
import './renderers/HBox';
import './renderers/VBox';
import './renderers/Image';
import './renderers/Images';
import './renderers/List';
import './renderers/Log';
import './renderers/Operation';
import './renderers/Page';
import './renderers/PaginationWrapper';
import './renderers/Panel';
import './renderers/Plain';
import './renderers/Property';
import './renderers/Portlet';
import './renderers/Spinner';
import './renderers/Table/index';
import './renderers/Tabs';
import './renderers/Tpl';
import './renderers/Mapping';
import './renderers/Progress';
import './renderers/Status';
import './renderers/Json';
import './renderers/Link';
import './renderers/Wizard';
import './renderers/Chart';
import './renderers/Container';
import './renderers/SwitchContainer';
import './renderers/SearchBox';
import './renderers/Service';
import './renderers/SparkLine';
import './renderers/Video';
import './renderers/Audio';
import './renderers/Nav';
import './renderers/Number';
import './renderers/Tasks';
import './renderers/Drawer';
import './renderers/Wrapper';
import './renderers/IFrame';
import './renderers/BarCode';
import './renderers/QRCode';
import './renderers/Icon';
import './renderers/Carousel';
import './renderers/AnchorNav';
import './renderers/Steps';
import './renderers/Timeline';
import './renderers/Markdown';
import './renderers/TableView';
import './renderers/Code';
import './renderers/WebComponent';
import './renderers/GridNav';
import './renderers/TooltipWrapper';
import './renderers/Tag';
import './renderers/Table2/index';
import './renderers/Words';
import './renderers/Password';
import './renderers/DateRange';
import './renderers/MultilineText';
import './renderers/OfficeViewer';
import './renderers/AMIS';

import './compat';
import './schemaExtend';
import type {
  BaseSchema,
  FormSchema,
  SchemaApi,
  SchemaCollection,
  SchemaExpression,
  SchemaObject
} from './Schema';
import type {TableViewSchema, TrObject} from './renderers/TableView';
import type {ActionSchema, ButtonSchema} from './renderers/Action';
import type {CRUDCommonSchema} from './renderers/CRUD';
import type {CRUD2Schema} from './renderers/CRUD2';
import type {TabsSchema} from './renderers/Tabs';
import {availableLanguages as EditorAvailableLanguages} from './renderers/Form/Editor';
import type {Action} from './types';
export * from './renderers/Form/IconPickerIcons';
export * from './renderers/Form/IconSelectStore';

export {
  BaseSchema,
  SchemaCollection,
  FormSchema,
  SchemaApi,
  SchemaObject,
  TableViewSchema,
  TrObject,
  ActionSchema,
  CRUDCommonSchema,
  ButtonSchema,
  CRUD2Schema,
  TabsSchema,
  SchemaExpression,
  Action,
  EditorAvailableLanguages
};
