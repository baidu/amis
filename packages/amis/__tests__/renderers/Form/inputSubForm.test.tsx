/**
 * 组件名称：InputSubForm 子表单
 * 单测内容：
 * 1. 基础点击勾选
 */

import React from 'react';
import {
  render,
  cleanup,
  screen,
  fireEvent,
  waitFor
} from '@testing-library/react';
import '../../../src';
import {render as amisRender} from '../../../src';
import {makeEnv, wait} from '../../helper';
import {clearStoresCache} from '../../../src';

afterEach(() => {
  cleanup();
  clearStoresCache();
});

test('Renderer:InputSubForm base', async () => {
  const onChange = jest.fn();
  const {getByText, container, baseElement} = render(
    amisRender({
      type: 'form',
      body: [
        {
          type: 'input-sub-form',
          name: 'form',
          label: '子Form',
          btnLabel: '设置子表单',
          onChange,
          form: {
            title: '配置子表单',
            body: [
              {
                name: 'a',
                label: 'A',
                type: 'input-text'
              },
              {
                name: 'b',
                label: 'B',
                type: 'input-text'
              }
            ]
          }
        }
      ]
    })
  );

  fireEvent.click(getByText('设置子表单'));

  expect(baseElement.querySelector('.cxd-Modal .cxd-Form')).toBeInTheDocument();

  const inputs = baseElement.querySelectorAll(
    '.cxd-Modal .cxd-Form .cxd-TextControl-input input'
  );
  expect(inputs!.length).toBe(2);
  expect(baseElement).toMatchSnapshot();

  fireEvent.change(inputs[0], {
    target: {value: 'text-one'}
  });
  await wait(200);
  fireEvent.change(inputs[1], {
    target: {value: 'text-two'}
  });

  await wait(200);
  fireEvent.click(getByText('确认'));
  await wait(500);

  expect(onChange).toBeCalled();
  expect(onChange.mock.calls[0][0]).toEqual({
    a: 'text-one',
    b: 'text-two'
  });
});

test('Renderer:InputSubForm with multiple & maxLength & btnLabel', async () => {
  const onChange = jest.fn();
  const {getByText, container, baseElement} = render(
    amisRender({
      type: 'form',
      body: [
        {
          type: 'input-sub-form',
          name: 'form',
          label: '子Form',
          btnLabel: '设置${a}',
          onChange,
          multiple: true,
          maxLength: 2,
          form: {
            title: '配置子表单',
            body: [
              {
                name: 'a',
                label: 'A',
                type: 'input-text'
              },
              {
                name: 'b',
                label: 'B',
                type: 'input-text'
              }
            ]
          }
        }
      ]
    })
  );

  async function addItem(val1: string, val2: string) {
    fireEvent.click(getByText('新增一项'));

    const inputs = baseElement.querySelectorAll(
      '.cxd-Modal .cxd-Form .cxd-TextControl-input input'
    );
    fireEvent.change(inputs[0], {
      target: {value: val1}
    });
    await wait(200);
    fireEvent.change(inputs[1], {
      target: {value: val2}
    });

    await wait(200);
    fireEvent.click(getByText('确认'));
    await wait(500);
  }

  await addItem('val-1-1', 'val-1-2');

  expect(onChange).toBeCalled();
  expect(onChange.mock.calls[0][0]).toEqual([
    {
      a: 'val-1-1',
      b: 'val-1-2'
    }
  ]);

  await addItem('val-2-1', 'val-2-2');
  expect(onChange).toBeCalledTimes(2);
  expect(onChange.mock.calls[1][0]).toEqual([
    {
      a: 'val-1-1',
      b: 'val-1-2'
    },
    {
      a: 'val-2-1',
      b: 'val-2-2'
    }
  ]);

  expect(
    container.querySelector('.cxd-SubForm-toolbar .cxd-SubForm-addBtn')!
  ).toHaveAttribute('disabled');

  const values = container.querySelectorAll(
    '.cxd-SubForm-values .cxd-SubForm-value'
  );

  expect(values!.length).toBe(2);
  expect(values[1]).toHaveTextContent('设置val-2-1');

  expect(container).toMatchSnapshot();
});

test('Renderer:InputSubForm with draggable & addable & removable', async () => {
  const onChange = jest.fn();
  const {getByText, container, baseElement} = render(
    amisRender({
      type: 'form',
      body: [
        {
          type: 'input-sub-form',
          name: 'form',
          label: '子Form',
          btnLabel: '设置${a}',
          onChange,
          multiple: true,
          draggable: true,
          addable: false,
          removable: true,
          form: {
            title: '配置子表单',
            body: [
              {
                name: 'a',
                label: 'A',
                type: 'input-text'
              },
              {
                name: 'b',
                label: 'B',
                type: 'input-text'
              }
            ]
          },
          value: [
            {
              a: 'val-a'
            },
            {
              a: 'val-b'
            },
            {
              a: 'val-c'
            },
            {
              a: 'val-d'
            }
          ]
        }
      ]
    })
  );

  let values = container.querySelectorAll(
    '.cxd-SubForm-values .cxd-SubForm-value'
  );

  expect(values!.length).toBe(4);
  expect(values[1]).toHaveTextContent('设置val-b');

  expect(
    values[0].querySelector('.cxd-SubForm-valueDragBar')
  ).toBeInTheDocument();

  fireEvent.click(values[0].querySelector('.cxd-SubForm-valueDel')!);

  await wait(200);

  values = container.querySelectorAll('.cxd-SubForm-values .cxd-SubForm-value');

  expect(values!.length).toBe(3);
  expect(values[1]).toHaveTextContent('设置val-c');

  fireEvent.click(values[0].querySelector('.cxd-SubForm-valueEdit')!);

  await wait(200);

  expect(baseElement.querySelector('.cxd-Modal .cxd-Form')).toBeInTheDocument();

  const firstInput = baseElement.querySelector(
    '.cxd-Modal .cxd-Form .cxd-TextControl-input input'
  )!;

  expect((firstInput as HTMLInputElement)!.value).toBe('val-b');

  fireEvent.change(firstInput, {
    target: {value: 'val-b-change'}
  });

  await wait(200);
  fireEvent.click(getByText('确认'));
  await wait(500);

  expect(onChange).toBeCalledTimes(2);
  expect(onChange.mock.calls[1][0]).toEqual([
    {
      a: 'val-b-change'
    },
    {
      a: 'val-c'
    },
    {
      a: 'val-d'
    }
  ]);
  expect(container).toMatchSnapshot();
});

test('Renderer:InputSubForm with addButtonClassName & itemsClassName & itemClassName & addButtonText & labelField', async () => {
  const {getByText, container, baseElement} = render(
    amisRender({
      type: 'form',
      body: [
        {
          type: 'input-sub-form',
          name: 'form2',
          label: '多选',
          multiple: true,
          labelField: 'title',
          itemsClassName: 'items-wrapper',
          itemClassName: 'item-classname',
          addButtonText: '自定义的新增',
          addButtonClassName: 'thisis-add-btn',
          value: [
            {
              title: 'val-a'
            },
            {
              title: 'val-b'
            }
          ],
          form: {
            title: '配置子表单',
            body: [
              {
                name: 'title',
                label: '标题',
                required: true,
                type: 'input-text'
              }
            ]
          }
        }
      ]
    })
  );

  await wait(500);

  expect(container).toMatchSnapshot();
  expect(container.querySelector('.cxd-SubForm-values')).toHaveClass(
    'items-wrapper'
  );
  expect(
    container.querySelector('.cxd-SubForm-values .cxd-SubForm-value')
  ).toHaveClass('item-classname');

  // labelField 这里不知为何不生效
  // expect(
  //   container.querySelector('.cxd-SubForm-values .cxd-SubForm-value')
  // ).toHaveTextContent('val-a');
  expect(container.querySelector('.cxd-SubForm-addBtn')).toHaveClass(
    'thisis-add-btn'
  );
  expect(container.querySelector('.cxd-SubForm-addBtn')).toHaveTextContent(
    '自定义的新增'
  );
});

test('Renderer:InputSubForm-can-access-superdata-1', async () => {
  const onChange = jest.fn();
  const {getByText, container, baseElement} = render(
    amisRender({
      type: 'form',
      data: {
        a: '123',
        b: '233'
      },
      body: [
        {
          type: 'input-sub-form',
          name: 'form',
          label: '子Form',
          btnLabel: '设置子表单',
          onChange,
          form: {
            title: '配置子表单',
            body: [
              {
                name: 'a',
                label: 'A',
                type: 'input-text'
              },
              {
                name: 'b',
                label: 'B',
                type: 'input-text'
              }
            ]
          }
        }
      ]
    })
  );

  fireEvent.click(getByText('设置子表单'));

  expect(baseElement.querySelector('.cxd-Modal .cxd-Form')).toBeInTheDocument();

  const inputs = baseElement.querySelectorAll(
    '.cxd-Modal .cxd-Form .cxd-TextControl-input input'
  );
  expect(inputs!.length).toBe(2);
  expect((inputs[0] as HTMLInputElement).value).toBe('');
  expect((inputs[1] as HTMLInputElement).value).toBe('');
});

test('Renderer:InputSubForm-can-access-superdata-2', async () => {
  const onChange = jest.fn();
  const {getByText, container, baseElement} = render(
    amisRender({
      type: 'form',
      data: {
        a: '123',
        b: '233'
      },
      body: [
        {
          type: 'input-sub-form',
          name: 'form',
          label: '子Form',
          btnLabel: '设置子表单',
          onChange,
          form: {
            canAccessSuperData: true,
            title: '配置子表单',
            body: [
              {
                name: 'a',
                label: 'A',
                type: 'input-text'
              },
              {
                name: 'b',
                label: 'B',
                type: 'input-text'
              }
            ]
          }
        }
      ]
    })
  );

  fireEvent.click(getByText('设置子表单'));

  expect(baseElement.querySelector('.cxd-Modal .cxd-Form')).toBeInTheDocument();

  const inputs = baseElement.querySelectorAll(
    '.cxd-Modal .cxd-Form .cxd-TextControl-input input'
  );
  expect(inputs!.length).toBe(2);
  expect((inputs[0] as HTMLInputElement).value).toBe('123');
  expect((inputs[1] as HTMLInputElement).value).toBe('233');
});
