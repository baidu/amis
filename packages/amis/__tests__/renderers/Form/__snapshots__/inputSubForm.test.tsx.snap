// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Renderer:InputSubForm base 1`] = `
<body
  class="is-modalOpened"
  style="width: calc(100% - 0px);"
>
  <div>
    <div
      class="cxd-Panel cxd-Panel--default cxd-Panel--form"
    >
      <div
        class="cxd-Panel-heading"
      >
        <h3
          class="cxd-Panel-title"
        >
          <span
            class="cxd-TplField"
          >
            <span>
              表单
            </span>
          </span>
        </h3>
      </div>
      <div
        class="cxd-Panel-body"
      >
        <form
          class="cxd-Form cxd-Form--normal"
          novalidate=""
        >
          <input
            style="display: none;"
            type="submit"
          />
          <div
            class="cxd-Form-item cxd-Form-item--normal"
            data-role="form-item"
          >
            <label
              class="cxd-Form-label"
            >
              <span>
                <span
                  class="cxd-TplField"
                >
                  <span>
                    子Form
                  </span>
                </span>
              </span>
            </label>
            <div
              class="cxd-SubFormControl cxd-Form-control"
            >
              <div
                class="cxd-SubForm-values"
              >
                <div
                  class="cxd-SubForm-value"
                  data-position="bottom"
                  data-tooltip="编辑详情"
                >
                  <span
                    class="cxd-SubForm-valueLabel"
                  >
                    <span
                      class="cxd-TplField"
                    >
                      <span>
                        设置子表单
                      </span>
                    </span>
                  </span>
                  <a
                    class="cxd-SubForm-valueEdit"
                  >
                    <icon-mock
                      classname="icon icon-pencil"
                      icon="pencil"
                    />
                  </a>
                </div>
              </div>
            </div>
          </div>
        </form>
      </div>
      <div
        class="cxd-Panel-footerWrap"
      >
        <div
          class="cxd-Panel-btnToolbar cxd-Panel-footer"
        >
          <button
            class="cxd-Button cxd-Button--primary cxd-Button--size-default"
            type="submit"
          >
            <span>
              提交
            </span>
          </button>
        </div>
      </div>
    </div>
  </div>
  <div
    class="amis-dialog-widget cxd-Modal cxd-Modal--1th"
    role="dialog"
  >
    <div
      class="cxd-Modal-overlay in"
    />
    <div
      class="cxd-Modal-content in"
    >
      <div
        class="cxd-Modal-header"
      >
        <a
          class="cxd-Modal-close"
          data-position="left"
          data-tooltip="关闭"
        >
          <icon-mock
            classname="icon icon-close"
            icon="close"
          />
        </a>
        <div
          class="cxd-Modal-title"
        >
          配置子表单
        </div>
      </div>
      <div
        class="cxd-Modal-body"
        role="dialog-body"
      >
        <div
          class="cxd-Form cxd-Form--horizontal"
          novalidate=""
        >
          <input
            style="display: none;"
            type="submit"
          />
          <div
            class="cxd-Form-item cxd-Form-item--horizontal"
            data-role="form-item"
          >
            <label
              class="cxd-Form-label cxd-Form-itemColumn--normal"
            >
              <span>
                <span
                  class="cxd-TplField"
                >
                  <span>
                    A
                  </span>
                </span>
              </span>
            </label>
            <div
              class="cxd-Form-value"
            >
              <div
                class="cxd-Form-control cxd-TextControl"
              >
                <div
                  class="cxd-TextControl-input"
                >
                  <input
                    autocomplete="off"
                    class=""
                    name="a"
                    placeholder=""
                    size="10"
                    type="text"
                    value=""
                  />
                </div>
              </div>
            </div>
          </div>
          <div
            class="cxd-Form-item cxd-Form-item--horizontal"
            data-role="form-item"
          >
            <label
              class="cxd-Form-label cxd-Form-itemColumn--normal"
            >
              <span>
                <span
                  class="cxd-TplField"
                >
                  <span>
                    B
                  </span>
                </span>
              </span>
            </label>
            <div
              class="cxd-Form-value"
            >
              <div
                class="cxd-Form-control cxd-TextControl"
              >
                <div
                  class="cxd-TextControl-input"
                >
                  <input
                    autocomplete="off"
                    class=""
                    name="b"
                    placeholder=""
                    size="10"
                    type="text"
                    value=""
                  />
                </div>
              </div>
            </div>
          </div>
        </div>
      </div>
      <div
        class="cxd-Modal-footer"
      >
        <button
          class="cxd-Button cxd-Button--default cxd-Button--size-default"
          type="button"
        >
          <span>
            取消
          </span>
        </button>
        <button
          class="cxd-Button cxd-Button--primary cxd-Button--size-default"
          type="button"
        >
          <span>
            确认
          </span>
        </button>
      </div>
    </div>
  </div>
</body>
`;

exports[`Renderer:InputSubForm with addButtonClassName & itemsClassName & itemClassName & addButtonText & labelField 1`] = `
<div>
  <div
    class="cxd-Panel cxd-Panel--default cxd-Panel--form"
  >
    <div
      class="cxd-Panel-heading"
    >
      <h3
        class="cxd-Panel-title"
      >
        <span
          class="cxd-TplField"
        >
          <span>
            表单
          </span>
        </span>
      </h3>
    </div>
    <div
      class="cxd-Panel-body"
    >
      <form
        class="cxd-Form cxd-Form--normal"
        novalidate=""
      >
        <input
          style="display: none;"
          type="submit"
        />
        <div
          class="cxd-Form-item cxd-Form-item--normal"
          data-role="form-item"
        >
          <label
            class="cxd-Form-label"
          >
            <span>
              <span
                class="cxd-TplField"
              >
                <span>
                  多选
                </span>
              </span>
            </span>
          </label>
          <div
            class="cxd-SubFormControl cxd-Form-control"
          >
            <div
              class="cxd-SubForm-values items-wrapper"
            >
              <div
                class="cxd-SubForm-value item-classname"
              >
                <span
                  class="cxd-SubForm-valueLabel"
                >
                  设置
                </span>
                <a
                  class="cxd-SubForm-valueEdit"
                  data-index="0"
                >
                  <icon-mock
                    classname="icon icon-pencil"
                    icon="pencil"
                  />
                </a>
                <a
                  class="cxd-SubForm-valueDel"
                  data-index="0"
                >
                  <icon-mock
                    classname="icon icon-close"
                    icon="close"
                  />
                </a>
              </div>
              <div
                class="cxd-SubForm-value item-classname"
              >
                <span
                  class="cxd-SubForm-valueLabel"
                >
                  设置
                </span>
                <a
                  class="cxd-SubForm-valueEdit"
                  data-index="1"
                >
                  <icon-mock
                    classname="icon icon-pencil"
                    icon="pencil"
                  />
                </a>
                <a
                  class="cxd-SubForm-valueDel"
                  data-index="1"
                >
                  <icon-mock
                    classname="icon icon-close"
                    icon="close"
                  />
                </a>
              </div>
            </div>
            <div
              class="cxd-SubForm-toolbar"
            >
              <button
                class="cxd-Button cxd-SubForm-addBtn thisis-add-btn"
                type="button"
              >
                <icon-mock
                  classname="icon icon-plus"
                  icon="plus"
                />
                <span>
                  自定义的新增
                </span>
              </button>
            </div>
          </div>
        </div>
      </form>
    </div>
    <div
      class="cxd-Panel-footerWrap"
    >
      <div
        class="cxd-Panel-btnToolbar cxd-Panel-footer"
      >
        <button
          class="cxd-Button cxd-Button--primary cxd-Button--size-default"
          type="submit"
        >
          <span>
            提交
          </span>
        </button>
      </div>
    </div>
  </div>
</div>
`;

exports[`Renderer:InputSubForm with draggable & addable & removable 1`] = `
<div>
  <div
    class="cxd-Panel cxd-Panel--default cxd-Panel--form"
  >
    <div
      class="cxd-Panel-heading"
    >
      <h3
        class="cxd-Panel-title"
      >
        <span
          class="cxd-TplField"
        >
          <span>
            表单
          </span>
        </span>
      </h3>
    </div>
    <div
      class="cxd-Panel-body"
    >
      <form
        class="cxd-Form cxd-Form--normal"
        novalidate=""
      >
        <input
          style="display: none;"
          type="submit"
        />
        <div
          class="cxd-Form-item cxd-Form-item--normal"
          data-role="form-item"
        >
          <label
            class="cxd-Form-label"
          >
            <span>
              <span
                class="cxd-TplField"
              >
                <span>
                  子Form
                </span>
              </span>
            </span>
          </label>
          <div
            class="cxd-SubFormControl cxd-Form-control"
          >
            <div
              class="cxd-SubForm-values"
            >
              <div
                class="cxd-SubForm-value"
              >
                <a
                  class="cxd-SubForm-valueDragBar"
                >
                  <icon-mock
                    classname="icon icon-drag-bar"
                    icon="drag-bar"
                  />
                </a>
                <span
                  class="cxd-SubForm-valueLabel"
                >
                  <span
                    class="cxd-TplField"
                  >
                    <span>
                      设置val-b-change
                    </span>
                  </span>
                </span>
                <a
                  class="cxd-SubForm-valueEdit"
                  data-index="0"
                >
                  <icon-mock
                    classname="icon icon-pencil"
                    icon="pencil"
                  />
                </a>
                <a
                  class="cxd-SubForm-valueDel"
                  data-index="0"
                >
                  <icon-mock
                    classname="icon icon-close"
                    icon="close"
                  />
                </a>
              </div>
              <div
                class="cxd-SubForm-value"
              >
                <a
                  class="cxd-SubForm-valueDragBar"
                >
                  <icon-mock
                    classname="icon icon-drag-bar"
                    icon="drag-bar"
                  />
                </a>
                <span
                  class="cxd-SubForm-valueLabel"
                >
                  <span
                    class="cxd-TplField"
                  >
                    <span>
                      设置val-c
                    </span>
                  </span>
                </span>
                <a
                  class="cxd-SubForm-valueEdit"
                  data-index="1"
                >
                  <icon-mock
                    classname="icon icon-pencil"
                    icon="pencil"
                  />
                </a>
                <a
                  class="cxd-SubForm-valueDel"
                  data-index="1"
                >
                  <icon-mock
                    classname="icon icon-close"
                    icon="close"
                  />
                </a>
              </div>
              <div
                class="cxd-SubForm-value"
              >
                <a
                  class="cxd-SubForm-valueDragBar"
                >
                  <icon-mock
                    classname="icon icon-drag-bar"
                    icon="drag-bar"
                  />
                </a>
                <span
                  class="cxd-SubForm-valueLabel"
                >
                  <span
                    class="cxd-TplField"
                  >
                    <span>
                      设置val-d
                    </span>
                  </span>
                </span>
                <a
                  class="cxd-SubForm-valueEdit"
                  data-index="2"
                >
                  <icon-mock
                    classname="icon icon-pencil"
                    icon="pencil"
                  />
                </a>
                <a
                  class="cxd-SubForm-valueDel"
                  data-index="2"
                >
                  <icon-mock
                    classname="icon icon-close"
                    icon="close"
                  />
                </a>
              </div>
            </div>
            <div
              class="cxd-SubForm-toolbar"
            >
              <span
                class="cxd-Combo-dragableTip"
              />
            </div>
          </div>
        </div>
      </form>
    </div>
    <div
      class="cxd-Panel-footerWrap"
    >
      <div
        class="cxd-Panel-btnToolbar cxd-Panel-footer"
      >
        <button
          class="cxd-Button cxd-Button--primary cxd-Button--size-default"
          type="submit"
        >
          <span>
            提交
          </span>
        </button>
      </div>
    </div>
  </div>
</div>
`;

exports[`Renderer:InputSubForm with multiple & maxLength & btnLabel 1`] = `
<div>
  <div
    class="cxd-Panel cxd-Panel--default cxd-Panel--form"
  >
    <div
      class="cxd-Panel-heading"
    >
      <h3
        class="cxd-Panel-title"
      >
        <span
          class="cxd-TplField"
        >
          <span>
            表单
          </span>
        </span>
      </h3>
    </div>
    <div
      class="cxd-Panel-body"
    >
      <form
        class="cxd-Form cxd-Form--normal"
        novalidate=""
      >
        <input
          style="display: none;"
          type="submit"
        />
        <div
          class="cxd-Form-item cxd-Form-item--normal"
          data-role="form-item"
        >
          <label
            class="cxd-Form-label"
          >
            <span>
              <span
                class="cxd-TplField"
              >
                <span>
                  子Form
                </span>
              </span>
            </span>
          </label>
          <div
            class="cxd-SubFormControl cxd-Form-control"
          >
            <div
              class="cxd-SubForm-values"
            >
              <div
                class="cxd-SubForm-value"
              >
                <span
                  class="cxd-SubForm-valueLabel"
                >
                  <span
                    class="cxd-TplField"
                  >
                    <span>
                      设置val-1-1
                    </span>
                  </span>
                </span>
                <a
                  class="cxd-SubForm-valueEdit"
                  data-index="0"
                >
                  <icon-mock
                    classname="icon icon-pencil"
                    icon="pencil"
                  />
                </a>
                <a
                  class="cxd-SubForm-valueDel"
                  data-index="0"
                >
                  <icon-mock
                    classname="icon icon-close"
                    icon="close"
                  />
                </a>
              </div>
              <div
                class="cxd-SubForm-value"
              >
                <span
                  class="cxd-SubForm-valueLabel"
                >
                  <span
                    class="cxd-TplField"
                  >
                    <span>
                      设置val-2-1
                    </span>
                  </span>
                </span>
                <a
                  class="cxd-SubForm-valueEdit"
                  data-index="1"
                >
                  <icon-mock
                    classname="icon icon-pencil"
                    icon="pencil"
                  />
                </a>
                <a
                  class="cxd-SubForm-valueDel"
                  data-index="1"
                >
                  <icon-mock
                    classname="icon icon-close"
                    icon="close"
                  />
                </a>
              </div>
            </div>
            <div
              class="cxd-SubForm-toolbar"
            >
              <button
                class="cxd-Button cxd-SubForm-addBtn"
                disabled=""
                type="button"
              >
                <icon-mock
                  classname="icon icon-plus"
                  icon="plus"
                />
                <span>
                  新增一项
                </span>
              </button>
            </div>
          </div>
        </div>
      </form>
    </div>
    <div
      class="cxd-Panel-footerWrap"
    >
      <div
        class="cxd-Panel-btnToolbar cxd-Panel-footer"
      >
        <button
          class="cxd-Button cxd-Button--primary cxd-Button--size-default"
          type="submit"
        >
          <span>
            提交
          </span>
        </button>
      </div>
    </div>
  </div>
</div>
`;
